Change from fetch to axios and refactor a bit

Aapzu opened the pull request:
This PR introduces changes, which make adding more domain objects and more API endpoint wrapper easier

- Remove `node-fetch` and introduce `axios`
  - Get rid of content parsing, error management and http/https agent hacks
- Move all API calls from domain object `Stream` to `StreamEndpoints.js`
- Some other, small refactorings

hpihkala commented on src/rest/StreamEndpoints.js:
```diff
@@ -1,45 +1,30 @@
     const url = `${this.options.restUrl}/streams/${streamId}`
-    const json = await authFetch(url, apiKey)
-    return json ? new Stream(this, json) : undefined
+    try {
```
Why this special handling for 404? I think the `authFetch` (old and new version) throws if anything goes wrong (including 404), and that's an ok pattern to keep.

Maybe the answer to "Should an exception be thrown if a resource is not found?" is a matter of opinion, and certainly depends on whether the user of the function expects the resource to be found. For `getStream` he certainly does, as he passes an unguessable id to it => not finding the stream would certainly be an unexpected error situation.

(In the old code, I'm not sure under what conditions `json` could be falsey. Maybe it's a mistake.)

Aapzu replied:
I guess I got confused because of the checking `return json ? new Stream(this, json) : undefined`

Aapzu replied:
Also got confused because inside `getOrCreateStream` the call to `getStream` is not inside `try ... catch` (`client.getOrCreateStream({ id: 'foo', name: 'foo' })` will now throw an error) https://github.com/streamr-dev/streamr-client/pull/18/files#diff-759bd3eca5b1471aff3290ec566e0ed2L74

hpihkala commented on src/rest/StreamEndpoints.js:
```diff
@@ -60,52 +45,70 @@ export async function createStream(props, apiKey = this.options.apiKey) {
-    } else if (props.name) {
-        json = await this.getStreamByName(props.name, apiKey)
-    }
+    let stream = await (props.id ? this.getStream(props.id, apiKey) : this.getStreamByName(props.name, apiKey))
```
Is this code really more readable?

Also it is not equivalent to the old code (new code queries `getStreamByName` even if `props.name` is not defined, which doesn't make sense).

hpihkala commented on src/rest/StreamEndpoints.js:
```diff
@@ -60,52 +45,70 @@ export async function createStream(props, apiKey = this.options.apiKey) {
-    } else {
-        return new Stream(this, json)
-    }
+    debug('Created stream: %s (%s)', props.name, stream.id)
```
This message should only be shown if a stream was really created (not fetched), just like before.

hpihkala commented on src/rest/StreamEndpoints.js:
```diff
@@ -60,52 +45,70 @@ export async function createStream(props, apiKey = this.options.apiKey) {
 
-    // If still nothing, throw
-    if (!json) {
-        throw new Error(`Unable to find or create stream: ${props.name}`)
```
Why not throw on "not found" anymore?

Aapzu replied:
`getStream` and `createStream` both throw errors if they do not succeed. If we have got this far we already know we got the response

hpihkala commented on src/rest/utils.js:
```diff
@@ -8,24 +8,18 @@ export const authFetch = async (url, apiKey, opts = {}) => {
+        const { data } = await axios(url, req)
+        return data
+    } catch (e) {
+        // TODO: what to log here?
```
What about this TODO? I recommend logging (using `debug`) exactly what happened, similar to the old error message of `Request to ${url} returned with error code ${res.status}: ${text}`

hpihkala commented on src/rest/utils.js:
```diff
@@ -8,24 +8,18 @@ export const authFetch = async (url, apiKey, opts = {}) => {
+        return data
+    } catch (e) {
+        // TODO: what to log here?
+        throw new Error(`${e.message} ${JSON.stringify(e.response.data)}`)
```
Definitely need to include the status code and the failed URL in the error message (like before).

Aapzu commented on src/rest/StreamEndpoints.js:
```diff
@@ -1,45 +1,30 @@
 
 const debug = debugFactory('StreamrClient')
 
-const agentSettings = {
```
Removed them because of a misunderstanding. Will add them back

timoxley commented on src/rest/StreamEndpoints.js:
```diff
@@ -60,52 +40,82 @@ export async function createStream(props, apiKey = this.options.apiKey) {
     let streamId
-    if (streamObjectOrId instanceof Stream) {
-        streamId = streamObjectOrId.id
+    if (streamOrId instanceof String || typeof streamOrId === 'string') {
```
Should this be `streamOrId instanceof Stream` rather than `String`?

timoxley replied:
Oh I see, yeah IMO I wouldn't even try to handle the `new String()` scenario, those types of values are going to be bad news for almost everything doing any kind of argument checks. `typeof` only IMO.

Aapzu replied:
Good to know, I just have a habit of checking a string that way (from some of the earliest "how to check if a value is a string in javascript" google searches)

timoxley commented on src/rest/domain/GenericDomainObject.js:
```diff
@@ -0,0 +1,13 @@
+        Object.assign(this, props)
+    }
+
+    setProps(props = {}) {
```
Maybe `setState` for parity with React's Component API?

Aapzu replied:
That's a good point. In a way I'd like to use some completely different word than `props` or `state`, but definitely from those two `state` is better

timoxley commented on src/rest/utils.js:
```diff
@@ -8,24 +8,17 @@ export const authFetch = async (url, apiKey, opts = {}) => {
+        const { data } = await axios(url, req)
+        return data
+    } catch (e) {
+        throw new Error(`Request to ${url} failed with http status code ${e.response.status}: ${JSON.stringify(e.response.data)}`)
```
This crash handler is going to crash if `e.response` or `e.response.data` don't exist for whatever reason.

Aapzu replied:
@hpihkala wanted the exception to show the url and the original exception doesn't show that

timoxley commented on test/integration/StreamrClient.spec.js:
```diff
@@ -22,16 +22,14 @@ describe('StreamrClient', function () {
     ])
         .catch((e) => {
-            if (e.errno === 'ENOTFOUND' || e.errno === 'ECONNREFUSED') {
+            if (e.code === 'ENOTFOUND' || e.code === 'ECONNREFUSED') {
```
are you sure these codes are right? when I run:

```
axios.get('http://httpstat.us/404').then(() => console.log('ok'), (e) => console.error(e.code)
```

the code is `undefined`.

Aapzu replied:
![](https://www.dropbox.com/s/8cwb7tk48mmctwe/Screenshot%202018-06-12%2011.58.52.png?raw=1)

At least when running the code in Node without `data-api` running in the background there is. But you're correct that when running in browser it's hard to get it to show the field (=didn't succeed)

Aapzu commented:
@hpihkala, added `httpAgent` and `httpsAgent` options back. Changed the error message of `authFetch` to include the url. Removed the special handling of 404 in `getStream`.
One more question: what do we want 
```
client.getOrCreateStream({
  id: 'foobar',
  name: 'name',
})
```
to do in a case where a Stream with id `foobar` does not exist?

hpihkala commented:
@Aapzu giving an `id` to `getOrCreateStream` doesn't make sense. This could be blocked already in the client by checking if the given object has an `id` field, and if it does, throwing an error. (The error could instruct the user to use `getStream` for getting streams by id).

hpihkala commented on src/rest/StreamEndpoints.js:
```diff
@@ -106,6 +108,12 @@ export function produceToStream(streamOrId, data, apiKey = this.options.apiKey,
         `${this.options.restUrl}/streams/${streamId}/data`,
         apiKey,
         {
+            httpAgent: http.Agent({
```
A few questions regarding this fix:

- Is this creating a new `Agent` for every request? (Here, actually **two** agents for every request). I would assume that the same `Agent` is meant to be shared across many requests.

- I'm not sure what exactly this line does. [The docs](https://nodejs.org/api/http.html#http_class_http_agent) show agent creation which includes the `new` keyword.

hpihkala commented:
Closing old unfinished PR
